core/src/main/scala/org/apache/spark/sql/SnappyContext.scala

```diff
@@ -546,41 +548,114 @@ class SnappyContext protected[spark](
+    * @param options Options for indexes. For e.g.
+    *                column table index - ("COLOCATE_WITH"->"CUSTOMER").
+    *                row table index - ("INDEX_TYPE"->"GLOBAL HASH") or ("INDEX_TYPE"->"UNIQUE")
+    */
+  def createIndex(indexName: String,
+                  baseTable: String,
+                  indexColumns: java.util.Map[String, Option[SortDirection]],
+                  options: java.util.Map[String, String]): Unit = {
+    createIndex(indexName, baseTable, indexColumns.asScala.toMap, options.asScala.toMap)
+  }
 
   /**
-   * Create Index on a SnappyData table (created by a call to createTable).
-   * @todo how can the user invoke this? sql?
-   */
-  private[sql] def createIndexOnTable(tableIdent: QualifiedTableName,
-      sql: String): Unit = {
-    //println("create-index" + " tablename=" + tableName    + " ,sql=" + sql)
+    * Create an index on a table.
+    * @param indexName Index name which goes in the catalog
+    * @param baseTable Fully qualified name of table on which the index is created.
+    * @param indexColumns Columns on which the index has to be created with the
+    *                     direction of sorting. Direction can be specified as None.
+    * @param options Options for indexes. For e.g.
+    *                column table index - ("COLOCATE_WITH"->"CUSTOMER").
+    *                row table index - ("INDEX_TYPE"->"GLOBAL HASH") or ("INDEX_TYPE"->"UNIQUE")
+    */
+  def createIndex(indexName: String,
+      baseTable: String,
+      indexColumns: Map[String, Option[SortDirection]],
+      options: Map[String, String]): Unit = {
+
+    val tableIdent = catalog.newQualifiedTableName(baseTable)
+    val indexIdent = catalog.newQualifiedTableName(indexName)
 
+    if (indexIdent.database != tableIdent.database) {
+      throw new AnalysisException(
+        s"Index and table have different databases " +
+          s"specified ${indexIdent.database} and ${tableIdent.database}")
+    }
     if (!catalog.tableExists(tableIdent)) {
       throw new AnalysisException(
-        s"$tableIdent is not an indexable table")
+        s"Could not find $tableIdent in catalog")
     }
-
-    //println("qualifiedTable=" + qualifiedTable)
+    import scala.collection.JavaConverters._
     catalog.lookupRelation(tableIdent) match {
-      case LogicalRelation(i: IndexableRelation, _) =>
-        i.createIndex(tableIdent.toString, sql)
+      case LogicalRelation(ir: IndexableRelation, _) =>
+        ir.createIndex(indexIdent,
+          tableIdent,
+          indexColumns,
+          options)
+
       case _ => throw new AnalysisException(
         s"$tableIdent is not an indexable table")
     }
   }
 
+  private[sql] def getIndexTable(indexIdent: QualifiedTableName): QualifiedTableName = {
+    new QualifiedTableName(Some(ColumnFormatRelation.INTERNAL_SCHEMA_NAME),
+      indexIdent.table)
+  }
+
+  private def constructDropSQL(indexName: String,
+              ifExists : Boolean): String = {
```

Comment:
After our conversation on phone, I checked both the Spark coding standards as well as existing Spark code. See: https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide -- clearly says 4 indent. Also looked at some methods like StaticMemoryManager.acquireStorageMemory which are using 4 indent only. I had created the original codeStyleSettings.xml looking at the guide so looks like that is the proper standard. Not clear what indentation Reynold's talked about.